pkg/server/api/entry.go

```diff
@@ -11,12 +11,130 @@ import (
 	"github.com/spiffe/spire/pkg/common/protoutil"
 	"github.com/spiffe/spire/pkg/common/x509util"
 	"github.com/spiffe/spire/proto/spire/common"
+	"google.golang.org/protobuf/proto"
 )
 
 const (
 	hintMaximumLength = 1024
 )
 
+type ReadOnlyEntry struct {
+	entry *types.Entry
+}
+
+func NewReadOnlyEntry(entry *types.Entry) ReadOnlyEntry {
+	return ReadOnlyEntry{
+		entry: entry,
+	}
+}
+
+func (e ReadOnlyEntry) GetId() string {
+	return e.entry.Id
+}
+
+func (e *ReadOnlyEntry) GetSpiffeId() *types.SPIFFEID {
+	return &types.SPIFFEID{
+		TrustDomain: e.entry.SpiffeId.TrustDomain,
+		Path:        e.entry.SpiffeId.Path,
+	}
+}
+
+func (e *ReadOnlyEntry) GetX509SvidTtl() int32 {
+	return e.entry.X509SvidTtl
+}
+
+func (e *ReadOnlyEntry) GetJwtSvidTtl() int32 {
+	return e.entry.JwtSvidTtl
+}
+
+func (e *ReadOnlyEntry) GetDnsNames() []string {
+	return slices.Clone(e.entry.DnsNames)
+}
+
+func (e *ReadOnlyEntry) GetRevisionNumber() int64 {
+	return e.entry.RevisionNumber
+}
+
+func (e *ReadOnlyEntry) GetCreatedAt() int64 {
+	return e.entry.CreatedAt
+}
+
+// Manually clone the entry instead of using the protobuf helpers
+// since those are two times slower.
+func (e *ReadOnlyEntry) Clone(mask *types.EntryMask) *types.Entry {
```

Comment:
I'm a bit concerned about the potential for field skew here. If the types.Entry struct is updated in the future (e.g., new fields are added), the manual Clone implementation will also need to be updated. Otherwise, newly added fields won't be cloned, which could lead to subtle, hard-to-detect bugs.

It might be helpful to add a comment warning about this, or ideally, include a test (perhaps using reflection) that ensures all fields are properly cloned.

Reply:
There's a test for this, which indeed uses reflection to make sure all fields are cloned: https://github.com/spiffe/spire/blob/d73d8a7b11c8152811cb4de843fe2173d55d0c16/pkg/server/api/entry_test.go#L727

Reply:
I missed that, thanks!